### PR TITLE
Delete EnumSet and remove its final use in rustc

### PR DESCRIPTION
**Pull request by jroesch**
This commit adds an implementation of `BuiltinBounds` that does not depend on external data structures and removes `EnumSet` in the standard collections library. We should probably have both a person from the compiler team and libs team look this one over.

r? @Gankro

### CODE REVIEW

**Comment by Gankro** [comment]
To be clear: EnumSet is already deprecated from libstd, and isn't re-exported there. As such you can just delete it. It's an internal implementation detail that was kept in exclusively for rustc's consumption.

File: src/librustc/middle/ty.rs [hunk 1]
```diff
@@ -2059,18 +2057,44 @@ pub struct ExistentialBounds<'tcx> {
...
 
 impl BuiltinBounds {
-       pub fn empty() -> BuiltinBounds {
-        BuiltinBounds(EnumSet::new())
+    pub fn empty() -> BuiltinBounds {
+        BuiltinBounds { bits: 0 }
+    }
+
+    pub fn insert(&mut self, bound: BuiltinBound) {
+        self.bits = match bound {
```
Comment (Gankro) [inline src/librustc/middle/ty.rs]: `self.bits |= match ...` ?

Comment (Gankro) [inline src/librustc/middle/ty.rs]: You could also just cast the BuiltinBounds to an integer. So

```rust
pub fn insert(&mut self, bound) {
    self.bits |= (1 << (bound as u8));
}

pub fn contains(&mut self, bound) {
    ((self.bits >> (bound as u8)) & 1) == 1
}
```

**Comment by Gankro** [comment]
The collection logic seems legit.

File: src/librustc/middle/ty.rs (hunk 1 above)
Comment (jroesch) [inline src/librustc/middle/ty.rs]: Yeah good point some of these are probably just my functional programmer side coming through ;)

**Comment by Gankro** [comment]
@jroesch What's the status on this; do you need any more review? r=me if this is good to go.

**Comment by jroesch** [comment]
@Gankro I didn't receive any negative feedback the second time so I assume that we are good to go.

**Comment by jroesch** [comment]
I didn't run the full test suite since this wasn't a huge change, one test was still mentioning CLike.

File: src/librustc_data_structures/bitset.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::ops::{Add, Sub, Shl, Shr, BitAnd, BitOr};
+use std::num::{Zero, One};
+use std::cmp::PartialEq;
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct BitSet<N> {
```
Comment (tamird) [inline src/librustc_data_structures/bitset.rs]: isn't this almost exactly the same as `bitflags`?

**Comment by steveklabnik** [comment]
@jroesch seems like a straightforward failure here

**Comment by tamird** [comment]
metaquestion: what is the difference between `EnumSet` and `bitflags`?

**Comment by arielb1** [comment]
@tamird

One is a macro which declares a statically-sized set, the other is a dynamically-resizing set.

**Comment by Gankro** [comment]
@arielb1 EnumSet also has a statically known size (number of variants).

**Comment by tamird** [comment]
Seems like we can avoid introducing `BitSet` and just use `bitflags!`

**Comment by Gankro** [comment]
This sound plausible, yes. @jroesch what's the status?

**Comment by jroesch** [comment]
@Gankro I had forgot this PR was still hanging around. I fixed the test failure, but we could just use `bitflags` if you don't foresee any problems.

**Comment by jroesch** [comment]
I've decided I'm just going to remove BuiltinBounds as per @eddyb 's suggestion will follow up on this PR.

**Comment by Gankro** [comment]
I'm just going to close this PR, since it sounds like the new one will just be fresh work. 

Good luck!

**Pull request closed without merging**